docs/specs/3-13-2023-commands-fit-n-finsh.md

```diff
@@ -0,0 +1,71 @@
+# 3/13/2023 - Commands Fit-and-Finish Proposal
+
+## Design
+
+Commands that interact with the core development workflow will be verb-based. This feels natural to the user as the user is building an application, provisioning infrastructure, and deploying their application to Azure -- which are all actions.
+
+For commands that are of configuration and management in-nature, it is fine to add a noun for a subcommand. For example, `azd env <action>` and `azd config <action>`.
+
+This means that the overall proposed design matches with what `azd` currently offers, and no changes need to occur.
+
+## `azd` usage output
+
+To help introduce user to the command layout, I propose slight changes to the current `azd` usage output (with UX input needed), that looks like the following:
+
+```
+Commands
+  Develop an app
+    init     	: Initialize a new application.
+    provision	: Provision the Azure resources for an application.
+    deploy   	: Deploy the application's code to Azure.
+    up       	: Initialize application, provision Azure resources, and deploy your project with a single command.
+
+  Monitor, test, and release your app
+    monitor  	: Monitor a deployed application.
+    pipeline 	: Manage and configure your deployment pipelines.
+
+  Authenticate with Azure
+    login    	: Log in to Azure.
+    logout   	: Log out of Azure.
+
+  Miscellaneous commands
+    config   	: Manage azd configurations (ex: default Azure subscription, location).
+    down     	: Delete Azure resources for an application.
+    env      	: Manage environments.
+    restore  	: Restore application dependencies.
+    template 	: Find and view template details.
+```
+
+The commands are grouped by their scenarios, with most important scenarios at the top. It is not in alphabetical order, rather the order the user is expected to run the commands. If we added `package` for example, would be after `init`.
+
+## Command changes
+
+The following changes are proposed:
+
+- `azd infra create` and `azd infra delete` should be removed, as the functionality already exists in `azd provision` and `azd down`, and users will find the duplication somewhat confusing.
+- The `--output` flag should be moved to `experimental` (hidden and unsupported unless with feature toggle), with the exception of `azd auth token --output json` which must be supported due to credential handlers.
+  - When the feature is hidden, the default output behavior of each command will apply.
+  - This is done as `--output` does not have a documented contract. It prints out `stdout` in JSON, while `stderr` messages also contain JSON-enveloped messages. We might consider splitting this behavior in the future.
+
+Discussion warranted:
+
+- Should we offer `azd restore`? I'm curious if users actually find this useful. We currently use this in our `tasks.json`, which seems useful, but the immediate block directly references `npm`, which defeats the nice abstraction. Excerpt from `tasks.json` below:
```

Comment:
`restore` looks like an odd thing because AZD today does not really have a CLI-based story for "run my app locally". Instead it relies on VS Code (tasks and debug configurations), which then leverage `restore`.  Moving `restore` functionality out of the AZD CLI is technically possible, but would complicate  VS Code tasks and debug configurations even more. Not sure if we want to go there. 

If we had `azd run` command for inner loop, we would probably discover that users do not want to restore their dependencies "just in case" every time they run the app, because it can take a significant amount of time. So in this case `restore` as a separate, user-controlled command is justified. Having either a `restore` command, or `--no-restore`/`--no-build` flag for `run` (or both) would address the need to iterate quickly. 

What I am trying to say is the fate of `restore` should be decided in the context of the question: "what should AZD local run/debug story look like, going forward?".

Reply:
I agree with this and think the build command along falls in the same category.

Reply:
Thank you for the broader thinking @karolz-ms, I agree we need to close down on inner-loop. Let's discuss [here](https://github.com/Azure/azure-dev/pull/1723/files#r1134729081).